### PR TITLE
Use explicit interval tree query for visibility test

### PR DESCRIPTION
**Pull request by rreusser**
This is a pretty major change so I don't expect a simple merge, but I thought I'd share my approach. I don't know if the swapping approach is the best since it'd be great to handle this on the objective c/gpu/whatnot level, but until then, I was trying to improve the robustness of the visibility check since it's been a challenge to optimize both robustness and performance (i.e. a small preemptiveLoading number results in cells being hidden accidentally; a large number slows it down a bit).

It wasn't quite clear to me what the built-in check was doing, so I thought I'd try maintaining an [interval tree](https://github.com/mikolalysenko/interval-tree-1d) explicitly. When a cell is laid out, it registers or updates its bounds in an interval tree. On each scroll event, the union of the previous visible range and current visible range is queried for intersecting cells in order to determine any visibility changes. I didn't like using onScroll, but a quick check indicates the check comes back under 1ms, so it shouldn't be a bottleneck at least.

It accepts a `preemptiveLoadDistance` that defines the distance in pixels outside the visible range at which cells become visible.

Final question: I had a hard time following the subtleties of the discussion at facebook/react-native#499 . Is there an approach yet that makes all of this obsolete?

### CODE REVIEW

**Comment by sghiassy** [comment]
Interval tree... very intriguing. I'll check out the PR tomorrow and get back to you.... thx for the PR!

**Comment by rreusser** [comment]
A fun weekend project, but I'm comparing this against the `removeClippedSubviews` approach that might be wayyyy better.

**Comment by sghiassy** [comment]
I originally tried using `removeClippedSubviews` but that still had `O(n)` memory growth in my testing. 

That was back in React Native `0.8` so they might've improved it by now

**Comment by rreusser** [comment]
Okay, this is not a very scientific test (should have profiled allocations), but I scrolled through an image feed for three minutes and here's what I have:

#### ListView without removedClippedSubviews

Memory increase linearly. Scrolling is _very_ choppy.
listview-no-remove-clipped

#### ListView with removeClippedSubviews

Uses `removeClippedSubviews` and `overflow:hidden` as described [here](https://github.com/facebook/react-native/issues/499#issuecomment-131583862). Memory increases linearly and is about the same as without `removeClippedSubviews` (I accidentally tapped the images a couple times and may not have scrolled at precisely the same rate so there's a margin of error). Scrolling is pretty smooth and when you snap to the top the image is present immediately.
listview-remove-clipped

#### SGListView with removeClippedSubviews

Memory increases but much less than without SGListView. I should have profiled allocations to see if it's the same as @sghiassy's test in the README. There's some flickering and stuttering because it's rerendering views but pretty smooth for realistic usage.
sglistview-remove-clipped

#### SGListView with interval tree and removeClippedSubviews

Basically the same as with @sghiassy's implementation except it maybe solves the issue where cells are accidentally removed even when they're onscreen.
interval-tree-remove-clipped-remove-clipped

**Comment by rreusser** [comment]
I guess I conclude from this that the memory grown isn't ideal, but maybe it was the lack of `removeClippedSubviews` that was causing it to be quite so janky. So maybe the answer is `removeClippedSubviews` if you're okay with some memory growth for ridiculously long scrolling sessions, and use SGListView if you really need to prevent that and a bit of rendering lag is okay with you—although I'm testing now to see if a much larger render-ahead distance helps that.

**Comment by rreusser** [comment]
Larger render-ahead distance indeed prevents seeing unrendered rows, but there'll always be swapping going on, so if your constraint is smoothness and memory growth is permissible, then ListView is the smoothest.

**Comment by rreusser** [comment]
While walking home I've realized that I've totally failed to handle re-layout correctly. All of the intervals need an update when any one changes. I can fix that. (Edit: need to check. But if things get an onLayout when they move, then this should be handled automagically. I'll confirm either way.)

**Comment by robertjpayne** [comment]
@rreusser in your tests did you ensure you added `overflow: 'hidden'` on each ListView row rather than the ListView itself?

Using react-native 0.11.0 I don't see as explosive of memory growth doing that as that ensures Image/Text nodes are destroyed.

**Comment by rreusser** [comment]
@robertjpayne thanks for the suggestion/critique! I'll check this tomorrow. I added it to the rows inside the SGListViewCells but not to the SGListViewCells where it may need to go. The memory management and rendering is moderately opaque to me so I appreciate any suggestions!

**Comment by rreusser** [comment]
Interesting. The discrepancy might only be in the profiling tool. `removeClippedSubviews` with a regular `ListView` shows linearly increasing memory pretty consistent with what's above (of the presumably bad sort). If I use the Instruments profiler though, the memory increases, but at a rate much more in line with the SGListView approach. Maybe the difference is an image/rendering cache that's not catastrophic. Hard to say. Just guesswork on my part, but maybe `removeClippedSubviews` is perfectly adequate on its own (much smoother, too).

**Comment by sghiassy** [comment]
@rreusser - Thanks again for the PR. 

I really like the interval-tree approach. It's the right data-structure for the job and am committed to bringing it into the repo if you are, but the current PR breaks things.

I didn't do a deep dive on what's breaking, but here's a debugger screenshot showing some early `NaN` values.

review

File: lib/SGListView.js [hunk 1]
```diff
@@ -61,10 +63,81 @@ var SGListView = React.createClass({
...
+    //                     y1p          y2p
+    //
+    //  current interval:           |------------|
+    //                             y1           y2
+    //
+    //  test interval:      |--------------------|
+    //                     yt1                  yt2
+    //
+    var scrollProperties = this.refs.nativeListView.scrollProperties
+    var y1 = scrollProperties.offset
```
Comment (sghiassy) [inline lib/SGListView.js:94]: `scrollProperties` doesn't have property `offset` Perhaps something changed between `0.8.0` and current?

File: lib/SGListView.js [hunk 2]
```diff
@@ -61,10 +63,81 @@ var SGListView = React.createClass({
...
+    y1 -= this.props.preemptiveLoadDistance
+    y2 += this.props.preemptiveLoadDistance
+
+    // See above illustration:
+    var yt1 = Math.min(this.y1p,y1)
+    var yt2 = Math.max(this.y2p,y2)
+
+    // Perform an interval tree query to find all cells that may become
+    // either hidden or visible:
+    this.layoutTree.queryInterval( yt1, yt2, (interval) => {
```
Comment (sghiassy) [inline lib/SGListView.js:109]: Won't this return multiple intervals? Don't we need to loop through them?

File: lib/SGListView.js [hunk 3]
```diff
@@ -61,10 +63,81 @@ var SGListView = React.createClass({
...
+      var cell = interval.cell
+
+      if( interval[1] < y1 || interval[0] > y2 ) {
```
Comment (sghiassy) [inline lib/SGListView.js]: Please reference the style guide. There's If/else spacing diversion as well as annotating logic tests.

File: lib/SGListView.js [hunk 4]
```diff
@@ -61,10 +63,81 @@ var SGListView = React.createClass({
...
+    var scrollProperties = this.refs.nativeListView.scrollProperties
+    var y1 = scrollProperties.offset
+    var y2 = y1 + scrollProperties.visibleLength
+
+    // Expand the interval so that things are rendered before they come on screen.
+    // Not DRY, but no need to overcomplicate this.
+    y1 -= this.props.preemptiveLoadDistance
+    y2 += this.props.preemptiveLoadDistance
+
+    return interval[0] < y2 && interval[1] > y1
```
Comment (sghiassy) [inline lib/SGListView.js]: Please annotate the logic test

File: lib/SGListViewCell.js [hunk 5]
```diff
@@ -50,20 +62,35 @@ var SGListViewCell = React.createClass({
...
+
+    this.props.addToTree(this.interval)
+
+    this.viewProperties.width = layout.width;
+    this.viewProperties.height = layout.height;
+
+    // Set the initial visibility state the first time it's laid out:
+    if( ! this.initialized ) {
+      this.initialized = true
+      if( ! this.props.isVisible(this.interval) ) this.setVisibility(false)
```
Comment (sghiassy) [inline lib/SGListViewCell.js]: This line is pretty terse. Can we update with the style guide: brackets, spacing and logic annotation.

File: lib/SGListView.js [hunk 6]
```diff
@@ -61,10 +63,81 @@ var SGListView = React.createClass({
...
+    //
+    //  current interval:           |------------|
+    //                             y1           y2
+    //
+    //  test interval:      |--------------------|
+    //                     yt1                  yt2
+    //
+    var scrollProperties = this.refs.nativeListView.scrollProperties
+    var y1 = scrollProperties.offset
+    var y2 = y1 + scrollProperties.visibleLength
```
Comment (sghiassy) [inline lib/SGListView.js:97]: Ditto with `visibleLength`.

File: lib/SGListView.js (hunk 1 above)
Comment (rreusser) [inline lib/SGListView.js:94]: Hmm. That may indeed be the case. I'm using react-native 0.11.2.

File: lib/SGListView.js (hunk 2 above)
Comment (rreusser) [inline lib/SGListView.js:109]: It does look suspicious at first, but the answer at mikolalysenko/interval-tree-1d#4 is that it's more performant to process the results as the tree is traversed than to store the results in a temporary array—which I'd imagine is pretty difficult to achieve without lots of unnecessary allocation and garbage collection.

File: lib/SGListView.js (hunk 3 above)
Comment (rreusser) [inline lib/SGListView.js]: Whoops! Yeah, it's right there in the style guide. Cleaning up now.

**Comment by rreusser** [comment]
It'll take a few minutes to create a new project with react native 0.8.0 and figure out what's breaking there, but I've committed code style and logic cleanup that should address the other issues.

Yeah, strictly speaking since the cells are adjacent and non-overlapping, this could all be accomplished more 'easily' with an ordered list of cell locations and a binary search, but since the interval tree is already implemented, this approach has the advantage of simplifying the difficult part to like two or three lines of code. :smile:

I should be able to give this a once-over for correctness and determine the 0.8 issue in the next day or two.

**Comment by rreusser** [comment]
Sorry I haven't made progress on this! We've been in the process of getting our thing out the door and `removeClippedSubviews` has basically worked perfectly! Would still like to finish this off in case anyone finds it useful though…

**Comment by rreusser** [comment]
Okay, I finally added a simple example. Works great, except the cells are initially invisible since they're rendered asynchronously. Works fine as soon as you scroll or if you use a settimeout before testing visibility. Appears though that I was a too confident. :frowning: 

screen shot 2015-11-24 at 11 29 31 am

**Comment by sghiassy** [comment]
So what's your thoughts on the PR? Want to close it out? Or is there a path forward?

**Comment by roysG** [comment]
Does it fix the problem for #21 ?
